crates/metaslang/bindings/src/builder/functions.rs

```diff
@@ -46,3 +54,50 @@ pub mod version {
         }
     }
 }
+
+mod resolver {
+    use std::sync::Arc;
+
+    use metaslang_cst::KindTypes;
+    use metaslang_graph_builder::functions::{Function, Functions, Parameters};
+    use metaslang_graph_builder::graph::{Graph, Value};
+    use metaslang_graph_builder::ExecutionError;
+
+    use crate::PathResolver;
+
+    pub fn add_functions<KT: KindTypes + 'static>(
+        functions: &mut Functions<KT>,
+        path_resolver: Arc<dyn PathResolver + Sync + Send>,
+    ) {
+        functions.add("resolve-path".into(), ResolvePath { path_resolver });
+    }
+
+    struct ResolvePath {
+        path_resolver: Arc<dyn PathResolver + Sync + Send>,
+    }
+
+    impl<KT: KindTypes> Function<KT> for ResolvePath {
+        fn call(
+            &self,
+            _graph: &mut Graph<KT>,
+            parameters: &mut dyn Parameters,
+        ) -> Result<Value, ExecutionError> {
+            let context_path = parameters.param()?.into_string()?;
+            let path_to_resolve = parameters.param()?.into_string()?;
+            parameters.finish()?;
+
+            if let Some(resolved_path) = self
+                .path_resolver
+                .as_ref()
+                .resolve_path(&context_path, &path_to_resolve)
+            {
+                Ok(resolved_path.into())
+            } else {
+                Err(ExecutionError::FunctionFailed(
```

Comment:
What would an `ExecutionError` entail? does it stop evaluating/building the graph? If a file cannot be resolved, can we still procceed, keeping the import unresolved?

There are many cases where this can happen, and we don't want to error out. For example, user is importing a contract from an NPM package, and they didn't run `npm install` yet, or they are running in a browser, where dependencies are not available.

Reply:
Yes, this would halt building the graph for this file. This branch will be taken if the resolver returns `None`. Currently the value returned by this function [is added as a push symbol attribute](https://github.com/NomicFoundation/slang/pull/1053/files/0fdb294b6197b2feb3fa9fa06386d0665927a593#diff-906114fedcbba5e9f9e302d77a57f998ecabef73d6cccdbdb1aaef8f3777bc16R162) to the import node.

In order to avoid halting execution, this needs to be a value that can be used as a symbol in the graph, ie. a string. Maybe we could use some sort of marker string like `UNRESOLVED IMPORT` instead? Do you think we should change the resolver interface as well?

Reply:
>  Maybe we could use some sort of marker string like UNRESOLVED IMPORT instead? 

Sure. Let's use something that won't ever be used by users. Maybe `__SLANG_UNRESOLVED_IMPORT_VALUE__` or something similar?

> Do you think we should change the resolver interface as well?

The resolver (implemented by the user) should be free to return `None` for many reasons. I think we should still try to provide partial results as best as we can.

Reply:
Makes sense. I'll make this change.